File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第5章：图像平滑/5.5/BilateralFiltering.cpp
#include<opencv2/core/core.hpp>
#include<opencv2/highgui/highgui.hpp>
#include<opencv2/imgproc/imgproc.hpp>
using namespace cv;
Mat getClosenessWeight(double sigma_g, Size size)
{
	//权重模板的宽高
	int W = size.width;
	int H = size.height;
	//模板的中心
	int cW = (W - 1) / 2;
	int cH = (H - 1) / 2;
	//权重模板
	Mat closenessWeight = Mat::zeros(size, CV_64FC1);
	for (int r = 0; r < H; r++)
	{
		for (int c = 0; c < W; c++)
		{
			double norm2 = pow(double(r - cH), 2.0) + pow(double(c - cW), 2.0);
			double sigma_g2 = 2.0*pow(sigma_g, 2.0);
			//赋值
			closenessWeight.at<double>(r, c) = exp(-norm2 / sigma_g2);
		}
	}
	return closenessWeight;
}
//双边滤波
Mat bfltGray(const Mat & image, Size winSize, float sigma_g, float sigma_d)
{
	int winH = winSize.height;
	int winW = winSize.width;
	//平滑窗口的长宽均是奇数
	CV_Assert(winH > 0 && winW > 0);
	CV_Assert(winH % 2 == 1 && winW % 2 == 1);
	if (winH == 1 && winW == 1)
		return image;
	int half_winW = (winW - 1) / 2;
	int half_winH = (winH - 1) / 2;
	//空间距离的权重因子
	Mat closenessWeight = getClosenessWeight(sigma_g, winSize);
	//图像的长宽
	int rows = image.rows;
	int cols = image.cols;
	//双边滤波后的输出图像
	Mat blfImage = Mat::ones(image.size(), CV_32FC1);
	//对每一个像素的领域进行核卷积
	for (int r = 0; r < rows; r++)
	{
		for (int c = 0; c < cols; c++)
		{
			double pixel = image.at<double>(r, c);
			//判断边界
			int rTop = (r - half_winH) < 0 ? 0 : r - half_winH;
			int rBottom = (r + half_winH) > rows - 1 ? rows - 1 : r + half_winH;
			int cLeft = (c - half_winW) < 0 ? 0 : c - half_winW;
			int cRight = (c + half_winW) > cols - 1 ? cols - 1 : c + half_winW;
			//核作用的区域
			Mat region = image(Rect(Point(cLeft, rTop), Point(cRight + 1, rBottom + 1))).clone();
			//相似性权重模板
			Mat similarityWeight;
			pow(region - pixel, 2.0, similarityWeight);
			exp(-0.5*similarityWeight / pow(sigma_d, 2), similarityWeight);
			similarityWeight /= pow(sigma_d, 2);
			//空间距离权重
			Rect regionRect = Rect(Point(cLeft - c + half_winW, rTop - r + half_winH), Point(cRight - c + half_winW + 1, rBottom - r + half_winH + 1));
			Mat closenessWeightTemp = closenessWeight(regionRect).clone();
			//两个权重模板点乘相乘，然后归一化
			Mat weightTemp = closenessWeightTemp.mul(similarityWeight);
			weightTemp = weightTemp / sum(weightTemp)[0];
			//权重模板与当前的领域对应位置相乘，然后求和
			Mat result = weightTemp.mul(region);
			blfImage.at<float>(r, c) = sum(result)[0];
		}
	}
	return blfImage;
}
int main(int argc, char*argv[])
{
	//输入图像
	Mat I = imread(argv[1], CV_LOAD_IMAGE_GRAYSCALE);
	if (!I.data)
		return -1;
	//灰度值归一化
	Mat fI;
	I.convertTo(fI, CV_64FC1, 1.0/ 255, 0);
	//双边滤波
	Mat blfI=bfltGray(fI, Size(33, 33), 19, 0.5);
	//显示原图和双边滤波的结果
	imshow("原图", I);
	imshow("双边滤波", blfI);
	//如果要保存为 8 位图,需要乘以255，并转换为 CV_8U
	blfI.convertTo(blfI, CV_8U, 255, 0);
	imwrite("blf.jpg", blfI);
	waitKey(0);
	return 0;
}

File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第5章：图像平滑/5.5/BilateralFiltering.py
# -*- coding: utf-8 -*-
import sys
import numpy as np
import cv2
import math
#基于空间距离的权重模板 ( 和计算高斯算子的过程是一样的 )
def getClosenessWeight(sigma_g,H,W):
    r,c = np.mgrid[0:H:1, 0:W:1]
    r -= (H-1)/2
    c -= (W-1)/2
    closeWeight = np.exp(-0.5*(np.power(r,2)+np.power(c,2))/math.pow(sigma_g,2))
    return closeWeight
# BilateralFiltering 双边滤波，返回的数据类型为浮点型
def bfltGray(I,H,W,sigma_g,sigma_d):
    #构建空间距离的权重模板
    closenessWeight = getClosenessWeight(sigma_g,H,W)
    #模板的中心点位置
    cH = (H -1)/2
    cW = (W -1)/2
    #图像矩阵的行数和列数
    rows,cols = I.shape
    #双边滤波后的结果
    bfltGrayImage = np.zeros(I.shape,np.float32)
    for r in xrange(rows):
        for c in xrange(cols):
            pixel = I[r][c]
            #判断边界
            rTop = 0 if r-cH < 0 else r-cH
            rBottom = rows-1 if r+cH > rows-1 else r+cH
            cLeft = 0 if c-cW < 0 else c-cW
            cRight = cols-1 if c+cW > cols-1 else c+cW
            #权重模板作用的区域
            region = I[rTop:rBottom+1,cLeft:cRight+1]
            #构建灰度值相似性的权重因子
            similarityWeightTemp = np.exp(-0.5*np.power(region -pixel,2.0)/math.pow(sigma_d,2))        
            closenessWeightTemp = closenessWeight[rTop-r+cH:rBottom-r+cH+1,cLeft-c+cW:cRight-c+cW+1]
            #两个权重模板相乘
            weightTemp = similarityWeightTemp*closenessWeightTemp
            weightTemp = weightTemp/np.sum(weightTemp)
            bfltGrayImage[r][c] = np.sum(region*weightTemp)
    return bfltGrayImage
#主函数
if __name__ =="__main__":
    if len(sys.argv)>1:
        image = cv2.imread(sys.argv[1],cv2.CV_LOAD_IMAGE_GRAYSCALE)
    else:
        print "Usge:python BFilter.py imageFile"
    #显示原图
    cv2.imshow("image",image)
    cv2.imwrite("image.png",image)
    #将灰度值归一化
    image = image/255.0
    #双边滤波
    bfltImage = bfltGray(image,33,33,10,0.8)
    #显示双边滤波的结果
    cv2.imshow("BilateralFiltering",bfltImage)
    bfltImage = bfltImage*255.0
    bfltImage = np.round(bfltImage)
    bfltImage = bfltImage.astype(np.uint8)
    #保存双边滤波的结果
    cv2.imwrite("BilateralFiltering.png",bfltImage)
    cv2.waitKey(0)
    cv2.destroyAllWindows()
